### PR TITLE
_linalg_qr_helper meta function

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #78836
* #78601
* #78604
* #78527
* __->__ #78603

Taken from https://github.com/albanD/subclass_zoo/blob/main/python_meta_tensor.py

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

**Review by lezcano — APPROVED** [review]
Just one check missing, otherwise lgtm

File: torch/_meta_registrations.py [hunk 1]
```diff
@@ -212,3 +212,34 @@ def meta_repeat_interleave_Tensor(repeats, output_size=None):
             "cannot repeat_interleave a meta tensor without output_size"
         )
     return repeats.new_empty(output_size)
+
+@torch.library.impl(meta_lib, "_linalg_qr_helper")
+def meta_linalg_qr_helper(input, mode):
```
[the remark this answers is not in the document]
Comment (Lezcano) [reply torch/_meta_registrations.py:217]: It's actually also missing the chekc on the dtype of `input`.

**Review by mruberry — APPROVED** [review]
Stamped

File: torch/_meta_registrations.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/_meta_registrations.py:217]: it doesn't look like the og function checks these

```
std::tuple<Tensor, Tensor> _linalg_qr_helper_default(const Tensor& input, c10::string_view mode) {
  bool compute_q, reduced_mode;
  std::tie(compute_q, reduced_mode) = _parse_qr_mode(mode);
  auto m = input.size(-2);
  auto n = input.size(-1);
  auto mn = std::min(m, n);

  // Allocate Q, R tensors with correct shape and memory layout
  Tensor Q;
  if (compute_q) {
    auto Qt_shape = input.sizes().vec();
    Qt_shape.end()[-2] = reduced_mode ? mn : m;
    Qt_shape.end()[-1] = m;
    Q = at::empty(Qt_shape, input.options());
    Q.transpose_(-2, -1); // make 'Q' with Fortran contiguous memory layout
  } else {
    Q = at::empty({0}, input.options());
  }

  auto Rt_shape = input.sizes().vec();
  Rt_shape.end()[-2] = n;
  Rt_shape.end()[-1] = (reduced_mode || !compute_q) ? mn : m;
  Tensor R = at::empty(Rt_shape, input.options());
  R.transpose_(-2, -1); // make 'R' with Fortran contiguous memory layout

  // Now fill Q, R tensors with the result
  linalg_qr_out_helper(input, Q, R, compute_q, reduced_mode);

  return std::make_tuple(Q, R);
}

```

[the remark this answers is not in the document]
Comment (ezyang) [reply torch/_meta_registrations.py:217]: alright i'm wrong it's in the out helper lol

File: torch/_meta_registrations.py [hunk 2]
```diff
@@ -212,3 +212,34 @@ def meta_repeat_interleave_Tensor(repeats, output_size=None):
...
+        reduced_mode = True
+    else:
+        raise RuntimeError(f"qr received unrecognized mode {mode}")
+    m = input.size(-2)
+    n = input.size(-1)
+    mn = min(m, n)
+    if compute_q:
+        Qt_shape = list(input.size())
+        Qt_shape[-2] = mn if reduced_mode else m
+        Qt_shape[-1] = m
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/_meta_registrations.py:235]: don't feel like golfing lol

File: torch/_meta_registrations.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Lezcano) [reply torch/_meta_registrations.py:217]: The preparation code for this function could be heavily simplified with Yukio's recent additions to the structured kernels logic. I should port it to those at some point.

**Pull request closed without merging**